packages/headless/src/controllers/url-manager/headless-url-manager.ts

```diff
@@ -0,0 +1,220 @@
+import {buildController, Controller} from '../controller/headless-controller';
+import {executeSearch} from '../../features/search/search-actions';
+import {ConfigurationSection} from '../../state/state-sections';
+import {logSearchboxSubmit} from '../../features/query/query-analytics-actions';
+import {
+  AnalyticsType,
+  makeNoopAnalyticsAction,
+} from '../../features/analytics/analytics-utils';
+import {logResultsSort} from '../../features/sort-criteria/sort-criteria-analytics-actions';
+import {
+  logFacetClearAll,
+  logFacetDeselect,
+  logFacetSelect,
+} from '../../features/facets/facet-set/facet-set-analytics-actions';
+import {buildSearchParameterSerializer} from '../../features/search-parameters/search-parameter-serializer';
+import {
+  buildSearchParameterManager,
+  getInitialSearchParameterState,
+} from '../search-parameter-manager/headless-search-parameter-manager';
+
+export interface UrlManagerProps {
+  /**
+   * The initial state that should be applied to the `UrlManager` controller.
+   */
+  initialState: UrlManagerInitialState;
+}
+
+export interface UrlManagerInitialState {
+  /**
+   * The part of the url containing the parameters affecting the search response.
+   */
+  fragment: string;
+}
+
+const initialStateSchema = new Schema<Required<UrlManagerInitialState>>({
+  fragment: new StringValue(),
+});
+
+/**
+ * The `UrlManager` controller allows managing an url which affect the search response.
+ * */
+export interface UrlManager extends Controller {
+  /**
+   * The state relevant to the `UrlManager` controller.
+   * */
+  state: UrlManagerState;
+  /**
+   * Updates the search parameters from the url & launches a search.
+   * @param fragment The part of the url containing the parameters affecting the search.
+   */
+  submitChanges(fragment: string): void;
+}
+
+export interface UrlManagerState {
+  /**
+   * The part of the url containing the parameters affecting the search.
+   */
+  url: string;
+}
+
+/**
+ * Creates a `UrlManager` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `UrlManager` properties.
+ * @returns A `UrlManager` controller instance.
+ */
+export function buildUrlManager(
+  engine: Engine<Partial<SearchParametersState> & ConfigurationSection>,
+  props: UrlManagerProps
+): UrlManager {
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+
+  validateInitialState(
+    engine,
+    initialStateSchema,
+    props.initialState,
+    'buildUrlManager'
+  );
+
+  const parameters = {
+    ...getInitialSearchParameterState(engine),
+    ...getSearchParameterStateFromFragment(props.initialState.fragment),
+  };
+  const searchParameterManager = buildSearchParameterManager(engine, {
+    initialState: {
+      parameters,
+    },
+  });
+
+  return {
+    ...controller,
+
+    get state() {
+      return {
+        url: buildSearchParameterSerializer().serialize(
```

Comment:
Since we are handling decoding the fragment, we should also handle encoding it here before exposing the string. 

One concern I have though is if `decodeURIComponent` and `encodeURIComponent` are not supported in some js run-time. I _think_ these functions are standard, and did not find evidence that they are missing.

The safer alternative would be to leave encoding/decoding up to the user of the controller. I would personally feel more comfortable if we did this, but I can see a pragmatic view to just do it since it _should_ work in > 99.9% of places.

Any thoughts? cc: @olamothe

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent

Reply:
It is a Standard built-in object it seems, I don't see why it would be unavailable in some runtime